In [3]:
import numpy as np

# Process attributes: [Arrival Time, Burst Time, Priority]
processes = [
    [0, 24, 3],
    [4, 3, 1],
    [5, 3, 4],
    [6, 12, 2]
]

# Time Quantum for Round Robin
time_quantum = 4

def calculate_waiting_turnaround_time(processes, completion_time):
    num_processes = len(processes)
    waiting_time = [0] * num_processes
    turnaround_time = [0] * num_processes

    for i in range(num_processes):
        turnaround_time[i] = completion_time[i] - processes[i][0]
        waiting_time[i] = turnaround_time[i] - processes[i][1]

    return waiting_time, turnaround_time

def fcfs_scheduling(processes):
    num_processes = len(processes)
    completion_time = [0] * num_processes
    current_time = 0

    for i in range(num_processes):
        current_time = max(current_time, processes[i][0])
        completion_time[i] = current_time + processes[i][1]
        current_time = completion_time[i]

    return completion_time

def sjf_scheduling(processes):
    num_processes = len(processes)
    completion_time = [0] * num_processes
    remaining_time = [processes[i][1] for i in range(num_processes)]
    current_time = 0

    while True:
        remaining_processes = [i for i in range(num_processes) if processes[i][0] <= current_time and remaining_time[i] > 0]
        if not remaining_processes:
            break

        shortest_process = min(remaining_processes, key=lambda x: remaining_time[x])
        current_time += 1
        remaining_time[shortest_process] -= 1

        if remaining_time[shortest_process] == 0:
            completion_time[shortest_process] = current_time

    return completion_time

def priority_scheduling(processes):
    num_processes = len(processes)
    completion_time = [0] * num_processes
    remaining_time = [processes[i][1] for i in range(num_processes)]
    current_time = 0

    while True:
        remaining_processes = [i for i in range(num_processes) if processes[i][0] <= current_time and remaining_time[i] > 0]
        if not remaining_processes:
            break

        highest_priority_process = min(remaining_processes, key=lambda x: processes[x][2])
        current_time += 1
        remaining_time[highest_priority_process] -= 1

        if remaining_time[highest_priority_process] == 0:
            completion_time[highest_priority_process] = current_time

    return completion_time

def round_robin_scheduling(processes, time_quantum):
    num_processes = len(processes)
    completion_time = [0] * num_processes
    remaining_time = [processes[i][1] for i in range(num_processes)]
    current_time = 0

    while True:
        all_processes_completed = True

        for i in range(num_processes):
            if processes[i][0] <= current_time and remaining_time[i] > 0:
                all_processes_completed = False
                if remaining_time[i] <= time_quantum:
                    current_time += remaining_time[i]
                    completion_time[i] = current_time
                    remaining_time[i] = 0
                else:
                    current_time += time_quantum
                    remaining_time[i] -= time_quantum

        if all_processes_completed:
            break

    return completion_time

# FCFS Scheduling
fcfs_completion_time = fcfs_scheduling(processes)
fcfs_waiting_time, fcfs_turnaround_time = calculate_waiting_turnaround_time(processes, fcfs_completion_time)

# SJF Scheduling
sjf_completion_time = sjf_scheduling(processes)
sjf_waiting_time, sjf_turnaround_time = calculate_waiting_turnaround_time(processes, sjf_completion_time)

# Priority Scheduling
priority_completion_time = priority_scheduling(processes)
priority_waiting_time, priority_turnaround_time = calculate_waiting_turnaround_time(processes, priority_completion_time)

# Round Robin Scheduling
rr_completion_time = round_robin_scheduling(processes, time_quantum)
rr_waiting_time, rr_turnaround_time = calculate_waiting_turnaround_time(processes, rr_completion_time)

# Calculate Average Waiting Time and Average Turnaround Time for each algorithm
avg_fcfs_waiting_time = np.mean(fcfs_waiting_time)
avg_fcfs_turnaround_time = np.mean(fcfs_turnaround_time)
avg_sjf_waiting_time = np.mean(sjf_waiting_time)
avg_sjf_turnaround_time = np.mean(sjf_turnaround_time)
avg_priority_waiting_time = np.mean(priority_waiting_time)
avg_priority_turnaround_time = np.mean(priority_turnaround_time)
avg_rr_waiting_time = np.mean(rr_waiting_time)
avg_rr_turnaround_time = np.mean(rr_turnaround_time)

# Print results
print("FCFS Completion Time:", fcfs_completion_time)
print("FCFS Waiting Time:", fcfs_waiting_time)
print("FCFS Turnaround Time:", fcfs_turnaround_time)
print("Average FCFS Waiting Time:", avg_fcfs_waiting_time)
print("Average FCFS Turnaround Time:", avg_fcfs_turnaround_time)

print("\nSJF Completion Time:", sjf_completion_time)
print("SJF Waiting Time:", sjf_waiting_time)
print("SJF Turnaround Time:", sjf_turnaround_time)
print("Average SJF Waiting Time:", avg_sjf_waiting_time)
print("Average SJF Turnaround Time:", avg_sjf_turnaround_time)

print("\nPriority Completion Time:", priority_completion_time)
print("Priority Waiting Time:", priority_waiting_time)
print("Priority Turnaround Time:", priority_turnaround_time)
print("Average Priority Waiting Time:", avg_priority_waiting_time)
print("Average Priority Turnaround Time:", avg_priority_turnaround_time)

print("\nRound Robin Completion Time:", rr_completion_time)
print("Round Robin Waiting Time:", rr_waiting_time)
print("Round Robin Turnaround Time:", rr_turnaround_time)
print("Average Round Robin Waiting Time:", avg_rr_waiting_time)
print("Average Round Robin Turnaround Time:", avg_rr_turnaround_time)


FCFS Completion Time: [24, 27, 30, 42]
FCFS Waiting Time: [0, 20, 22, 24]
FCFS Turnaround Time: [24, 23, 25, 36]
Average FCFS Waiting Time: 16.5
Average FCFS Turnaround Time: 27.0

SJF Completion Time: [42, 7, 10, 22]
SJF Waiting Time: [18, 0, 2, 4]
SJF Turnaround Time: [42, 3, 5, 16]
Average SJF Waiting Time: 6.0
Average SJF Turnaround Time: 16.5

Priority Completion Time: [39, 7, 42, 19]
Priority Waiting Time: [15, 0, 34, 1]
Priority Turnaround Time: [39, 3, 37, 13]
Average Priority Waiting Time: 12.5
Average Priority Turnaround Time: 23.0

Round Robin Completion Time: [42, 7, 10, 30]
Round Robin Waiting Time: [18, 0, 2, 12]
Round Robin Turnaround Time: [42, 3, 5, 24]
Average Round Robin Waiting Time: 8.0
Average Round Robin Turnaround Time: 18.5


FCFS (First-Come-First-Serve):

Average Waiting Time: Calculated as 16.5
Average Turnaround Time: Calculated as 27.0
    
SJF (Shortest Job First):

Average Waiting Time: Calculated as 6.0
Average Turnaround Time: Calculated as 16.5
    
Priority Scheduling:

Average Waiting Time: Calculated as 12.5
Average Turnaround Time: Calculated as 23.0
    
Round Robin (RR) with Time Quantum 4:

Average Waiting Time: Calculated as 8.0
Average Turnaround Time: Calculated as 18.5
Analysis:

SJF (Shortest Job First) performs the best with the lowest average waiting time and average turnaround time, making it the most suitable algorithm for this set of processes.
Round Robin (RR) with a time quantum of 4 provides a fair balance between waiting time and turnaround time, making it a reasonable choice if fairness is a priority.
Priority Scheduling performs better than FCFS but is outperformed by SJF and Round Robin in this scenario.
FCFS (First-Come-First-Serve) performs the worst in terms of both average waiting time and average turnaround time.